### ISSUE TITLE
[Ch 21][Notebook][Minor] 예상 소요 시간이 실측의 12~50배 + FAQ Q8 표 시간·정확도 순서 역전

### ISSUE THREAD

**Issue by yum-yeom** [yoon-gu/neuqes-101#127]
## 대상

- **챕터**: Ch 21
- **산출물**: Notebook (`21_en_bert_classify.ipynb` 머리말·§변경점 비교표·FAQ Q8) + `21_en_bert_classify/README.md` — 전파: `pages/21-*.md`, `book/chapters/ch21_en_bert_classify.tex`
- **기준 커밋**: `c16215e`
- **환경**: 실행본 `executed/21_en_bert_classify.ipynb`(`c16215e`, Colab T4, 129.2초) 대조. **Colab 재실행 미수행**

## 문제 위치

**결함 위치**

| 파일 | 위치 | 링크 | 내용 |
|---|---|---|---|
| `21_en_bert_classify/21_en_bert_classify.ipynb` | cell[0] (첫 셀) — **예상 소요 시간** 문단 | [L3](https://github.com/yoon-gu/neuqes-101/blob/c16215e/21_en_bert_classify/21_en_bert_classify.ipynb?plain=1#L3) | 약 25-28분 |
| `21_en_bert_classify/21_en_bert_classify.ipynb` | cell[2] — `## 🔄 변경점` 안 Ch 10 비교표 | [L15](https://github.com/yoon-gu/neuqes-101/blob/c16215e/21_en_bert_classify/21_en_bert_classify.ipynb?plain=1#L15) | 사전학습 시간 'T4 약 10분' |
| `21_en_bert_classify/21_en_bert_classify.ipynb` | cell[47] — `## ❓ FAQ` Q8 크기별 표 | [L333](https://github.com/yoon-gu/neuqes-101/blob/c16215e/21_en_bert_classify/21_en_bert_classify.ipynb?plain=1#L333) | 시간·accuracy 순서 역전 |
| `21_en_bert_classify/README.md` | L31 / L34 / L54 | [L34](https://github.com/yoon-gu/neuqes-101/blob/c16215e/21_en_bert_classify/README.md?plain=1#L34) | MLM 1 epoch·25-28분·10-12분 |

**근거 (실행본)**

| 파일 | 위치 | 링크 | 내용 |
|---|---|---|---|
| `executed/21_en_bert_classify.ipynb` | cell[23] — MLM 학습 | [열기](https://github.com/yoon-gu/neuqes-101/blob/c16215e/executed/21_en_bert_classify.ipynb) | `done in 0.2 min`, `[198/198 00:13]` |
| `executed/21_en_bert_classify.ipynb` | cell[31] — 분류 학습 | [열기](https://github.com/yoon-gu/neuqes-101/blob/c16215e/executed/21_en_bert_classify.ipynb) | `done in 0.2 min`, `[626/626 00:13]` |
| `executed/21_en_bert_classify.ipynb` | metadata.executed_from | [열기](https://github.com/yoon-gu/neuqes-101/blob/c16215e/executed/21_en_bert_classify.ipynb) | elapsed_sec 129.2 |
## 실제 결과

예상 소요 시간이 실측의 **12~50배** 입니다.

| 자리 | 본문 표기 | 실측 |
|---|---|---|
| 노트북 머리말 | 약 25-28분 (다운로드 2분 + MLM 3 epoch **8-10분** + 분류 2 epoch **8-10분** + 평가 2분) | **전체 2분 9초** (`executed_from.elapsed_sec = 129.2`) |
| §변경점 비교표 | 사전학습 시간 "T4 약 10분" | MLM **0.2분** — `MLM pretraining done in 0.2 min`, 진행표 `[198/198 00:13]` |
| `pages`·tex 비교표 | "T4 약 10-12분" | 〃 |
| README §환경 | 약 25-28분 (**MLM 1 epoch** 약 10-12분 + 분류 8-10분 …) | 분류 fine-tune **0.2분** — `[626/626 00:13]` |

README 는 "MLM 1 epoch" 이라 코드(`MLM_EPOCHS = 3`)와도 어긋납니다 (별건 이슈로도 등록).

FAQ Q8 의 크기별 표도 같은 뿌리로 어긋나 있습니다.

| 모델 크기 | 표의 학습 시간 | 표의 예상 accuracy |
|---|---|---|
| hidden=128, layer=2 (약 5M) | 약 5분 | 65-72% |
| **hidden=256, layer=4 (이번 챕터, 약 10M)** | **약 1분** | 실행본 참조 (실측 0.647) |
| hidden=384, layer=6 (약 20M) | 약 30분 | 78-88% |

작은 모델이 이번 챕터보다 **5배 오래 걸리는** 순서이고, 예상 accuracy 도 5M(65-72%)이 이번 챕터 실측(0.647)보다 높게 적혀 있습니다.

## 기대 결과

시간 추정이 실측 규모와 같은 자릿수여야 하고, 노트북·README·FAQ 표가 서로 어긋나지 않아야 합니다.

## 수정 제안

- 머리말·비교표는 실측 기준으로 (전체 약 2-3분, MLM·분류 각 약 15초). 데이터 다운로드가 대부분을 차지하므로 "대부분 다운로드 시간" 이라고 밝히면 학습자가 덜 당황합니다.
- README 는 노트북과 같은 값으로 통일하고 "MLM 1 epoch" → 3 epoch.
- FAQ Q8 표는 시간·accuracy 열 모두 단조성이 깨져 있으니, 이번 챕터 행을 실측으로 고정한 뒤 나머지 행을 그에 맞춰 다시 잡는 편이 낫습니다.

## 재현 절차

```bash
python3 -c "
import json; nb=json.load(open('executed/21_en_bert_classify.ipynb'))
print(nb['metadata']['executed_from'])
for c in nb['cells']:
    for o in c.get('outputs',[]):
        if o.get('output_type')=='stream':
            for l in ''.join(o['text']).split('\n'):
                if 'done in' in l: print(l.strip())
"
grep -n '25-28분\|10-12분\|약 10분' 21_en_bert_classify/README.md 21_en_bert_classify/21_en_bert_classify.ipynb
```

## 심각도

**Major** — 학습자가 30분짜리 작업으로 알고 일정을 잡습니다. Ch 20 [#120](https://github.com/yoon-gu/neuqes-101/issues/120) 과 같은 유형입니다.

## 담당

- 등록자·재검수자: @yum-yeom (페어 C)
- 담당: @fluentmin

## 연결

- 챕터 검수 이슈: [#45](https://github.com/yoon-gu/neuqes-101/issues/45)
- 같은 유형 선례: [#120](https://github.com/yoon-gu/neuqes-101/issues/120) (Ch 20), [#102](https://github.com/yoon-gu/neuqes-101/issues/102)·[#109](https://github.com/yoon-gu/neuqes-101/issues/109) (Ch 12·13)

**Comment by yum-yeom** [comment]
## 수정했습니다 — 실측으로 정정

브랜치 `fix/ch21-review-2026-09` · 커밋 [`b3a5045`](https://github.com/yoon-gu/neuqes-101/commit/b3a5045)

기준은 실행본 `executed/21_en_bert_classify.ipynb` (`c16215e`, Colab T4) 입니다 — 전체 129.2초, MLM `[198/198 00:13]`, 분류 `[626/626 00:13]`.

### 1. 노트북 머리말

```diff
- **예상 소요 시간**: 약 25-28분 (Wikitext-103 다운로드·필터링 약 2분 + MLM 3 epoch 약 8-10분 + 분류 fine-tune 2 epoch 약 8-10분 + 평가 약 2분)
+ **예상 소요 시간**: 약 3-5분 — 대부분이 데이터 다운로드입니다 (실행본 기준 전체 2분 9초: 다운로드·전처리 약 1분 30초 + MLM 3 epoch 약 13초 + 분류 fine-tune 2 epoch 약 13초 + 평가·시각화 수 초). 다운로드 속도에 따라 달라집니다.
```

"대부분이 다운로드" 를 밝혀, 학습이 순식간에 끝나는 걸 보고 당황하지 않게 했습니다.

### 2. §변경점 비교표 · README

```diff
- | 사전학습 시간 | TPU 수일 (대규모 인프라) | **T4 약 10분** | |
+ | 사전학습 시간 | TPU 수일 (대규모 인프라) | **T4 약 13초** (2K × 3 epoch = 198 step) | |
```

README 는 §환경(25-28분·MLM 1 epoch 10-12분)과 비교표(10-12분) 두 곳을 노트북과 같은 값으로 통일했습니다. "MLM 1 epoch" 표기는 #129 에서 함께 정리했습니다.

### 3. FAQ Q8 크기별 표

작은 모델이 더 오래 걸리는 순서였고, 예상 accuracy 도 5M 행(65-72%)이 이번 챕터 실측(0.647)보다 높았습니다. 이번 챕터 행을 실측으로 고정하고 나머지를 그에 맞춰 다시 잡았습니다.

```diff
- | hidden=128, layer=2 | 약 5M | 약 5분 | 65-72% |
- | **hidden=256, layer=4 (이번 챕터)** | **약 10M** | **약 1분** | **실행본 참조 (§6)** |
- | hidden=384, layer=6 | 약 20M | 약 30분 | 78-88% (T4 30분 한계) |
+ | hidden=128, layer=2 | 약 5M | 약 20초 | 이번 챕터보다 낮음 |
+ | **hidden=256, layer=4 (이번 챕터)** | **약 10M** | **약 30초** (실측 — MLM 13초 + 분류 13초) | **실행본 참조 (§6)** |
+ | hidden=384, layer=6 | 약 20M | 약 1분 | 조금 높음 |
```

**측정하지 않은 행은 여전히 추정치입니다.** 실측한 이번 챕터 행에서 파라미터 수로 대략 비례시킨 값이고, 표 헤더와 아래 주석에 *이 챕터 데이터 기준* 임을 밝혔습니다. accuracy 열도 지어낸 구간 대신 이번 챕터 대비 상대 서술로 바꿨습니다. 근거 없는 절대 수치를 다시 심고 싶지 않아서입니다.

FAQ Q3 의 "`N_MLM_TRAIN = 20000` + `MLM_EPOCHS = 1` 정도가 T4 30분 룰 안에서 최대치" 도 실측 기준으로 다시 썼습니다 — 30분 룰에 먼저 걸리는 쪽은 학습이 아니라 다운로드·전처리입니다.

### 남은 것

코드 변경이 없어 재실행은 하지 않았고, 실행본은 마크다운만 동기화했습니다([`9c32ad2`](https://github.com/yoon-gu/neuqes-101/commit/9c32ad2)). WikiDocs·tex·EPUB 재생성은 #45 에서 이어갑니다.

@fluentmin 확인 후 닫아 주세요.

**Comment by yum-yeom** [comment]
## 정정 — 기준 브랜치와 #123 상태

등록할 때 쓴 기준이 틀렸습니다. 본문을 조용히 고치지 않고 여기 적습니다.

**1. 기준 브랜치·커밋**

`fix/ch21-review-2026-09` @ `c16215e` 로 적었는데, 팀이 통일한 Ch 21 브랜치는 **`ch-21/review-2026-09`** (Ch 20 의 `ch-20/review-2026-09` 와 같은 규칙) 이고 커밋 2개가 더 있습니다.

- `428f2c2` — CLAUDE.md 에 검수 기준 논의 #63 합의 4건 반영
- `8e65def` — Ch 21 모델 random init 시드 (#123)

**이 이슈의 결함은 새 브랜치(`8e65def`)에서도 그대로 재현됨을 확인했습니다.** 지적 내용에는 영향이 없습니다.

**2. #123 상태**

본문·앞선 댓글에서 "본편은 시드가 걸려 있지 않다" 고 적었는데, 새 브랜치에서는 `8e65def` 로 **이미 걸려 있습니다** (셀 15·27 모델 생성 직전 `set_seed(SEED)`). 제가 옛 브랜치만 보고 쓴 것입니다.

다만 그 커밋이 코드셀 3개(7·15·27)를 바꾼 뒤 **재실행이 되지 않아** 실행본은 8/13 판 그대로입니다 (`executed_from.executed_at = 2026-08-13`, `source_sha256` 불일치). #123 의 Ch 21 몫은 **재실행이 남아 있습니다.**

**3. 수정 커밋 위치**

앞 댓글의 `b3a5045` · `9c32ad2` 는 옛 브랜치 커밋입니다. 새 브랜치로 옮겨 다시 올렸습니다.

- [`ec8e175`](https://github.com/yoon-gu/neuqes-101/commit/ec8e175) — 서술 수정 (내용 동일)
- [`4bae94f`](https://github.com/yoon-gu/neuqes-101/commit/4bae94f) — 실행본 마크다운 동기화. `source_sha256` 은 실제 실행된 소스(`c16215e`) 값을 그대로 두고 `executed_from.note` 에 사유를 남겼습니다 — 지금의 해시 불일치가 **#123 재실행 필요** 를 뜻하도록.


**4. 심각도를 Major → Minor 로 낮췄습니다**

검수 스킬(`.claude/skills/chapter-review/SKILL.md`, `fix/ch12-review-2026-09` 브랜치)의 판정표가 이 유형을 이렇게 못박고 있습니다.

> \"약 12분\" 인데 실측 37초 → **결함 아님** — 런타임·네트워크에 좌우됨

같은 유형인 Ch 12 [#102](https://github.com/yoon-gu/neuqes-101/issues/102) · Ch 13 [#109](https://github.com/yoon-gu/neuqes-101/issues/109) 도 Minor 입니다. 등급을 맞추는 게 맞다고 보아 라벨과 제목을 Minor 로 바꾸고 `type:runtime` 을 추가했습니다.

다만 이 이슈에는 시간 변동으로 설명되지 않는 것이 함께 들어 있어 닫지는 않았습니다 — **FAQ Q8 표의 순서 역전**(작은 모델이 5배 느리고, 5M 행의 예상 accuracy 가 이번 챕터 실측보다 높음)은 재실행해도 사라지지 않는 내부 모순입니다. 제목에 그 부분을 드러냈습니다.

[#109](https://github.com/yoon-gu/neuqes-101/issues/109) 가 제안한 대로 **전 챕터 표기 방식을 한 번에 정하는 편**이 낫다고 보며, 그 결정이 [#63](https://github.com/yoon-gu/neuqes-101/issues/63) 에서 나오면 이 이슈도 거기에 맞추겠습니다.

## 전파 범위

- `pages/21-en_bert_classify.md:9` (25-28분) · `:73` (T4 약 10분)
- `pages/21-en_bert_classify-anatomy.md:155` (T4 약 10-12분)
- `pages/21-en_bert_classify-wrapup.md:145` (Q8 표)
- `book/chapters/ch21_en_bert_classify.tex:170` · `:1638`

## 재실행 필요 여부

**불필요** — 마크다운·README 만 고쳤습니다.

**Comment by fluentmin** [comment]
확인했습니다 (@fluentmin, `ch-21/review-2026-09` @ `7e45c3f`) — 남아 있던 두 곳을 마저 고치고 재실행으로 검증했습니다.

### 먼저 고쳐진 자리 — 실행본과 대조해 확인

- 노트북 머리말 `약 25-28분` → 실측 기준 + "대부분이 데이터 다운로드" ✅
- §변경점 비교표 `T4 약 10분` → 실측 ✅
- README §환경 · 비교표 — 노트북과 통일 ✅
- FAQ Q8 크기별 표 — 순서 역전 해소, accuracy 열을 이번 챕터 대비 상대 서술로 ✅

측정하지 않은 행을 지어낸 절대 수치 대신 상대 서술로 바꾼 판단에 동의합니다. Major → Minor 재조정도 #102·#109 와 등급이 맞습니다.

### 남아 있던 두 곳 ([`2237b28`](https://github.com/yoon-gu/neuqes-101/commit/2237b28))

챕터 안에 *사전학습 시간* 행이 셋인데 §변경점 하나만 고쳐져 있었습니다.

```
cell[2]:  T4 약 13초   ← 고쳐진 자리
cell[40]: T4 약 10-12분  (🔬 해부 — Ch 10 비교표)
cell[47]: T4 10분        (FAQ Q7 — DistilBERT 비교표)
```

이슈 §전파 범위에 적어 주신 `pages/21-en_bert_classify-anatomy.md:155` · `tex:1638` 이 cell[40], `tex:1928` 이 cell[47] 입니다. **전파본이 아니라 노트북이 출처인 자리** 라 #45 에서 재변환해도 `10-12분` 이 그대로 다시 생성될 상태였습니다. 같은 챕터에서 `13초` 와 `10-12분` 이 나란히 서 있었습니다.

### 재실행하니 13초가 또 어긋났습니다 ([`4d3e9ae`](https://github.com/yoon-gu/neuqes-101/commit/4d3e9ae))

수정한 노트북을 Colab T4 에서 다시 돌렸더니 MLM·분류가 각각 **14초**, 전체 **2분 13초** 였습니다. 맞춰 놓은 `13초` 가 한 판 만에 틀린 것입니다.

값을 계속 따라가면 [#72](https://github.com/yoon-gu/neuqes-101/issues/72) 가 겪은 churn 이 시간 축에서 반복됩니다. 그래서 두 판(13초·14초)을 모두 품는 **`약 15초`** 로 올렸습니다. 이 표가 말하려는 건 *TPU 수일* 과 *T4 십몇 초* 의 대비라, 이 자릿수면 충분하고 다음 실행에서도 잘 안 틀립니다. 전체 소요 시간은 성격이 달라(측정 기록) 실행본 기준 `2분 13초` 로 갱신하고 출처를 함께 적었습니다.

결과적으로 챕터 안 시간 표기 **8곳** 이 한 값으로 정리됐습니다 — 노트북 표 3곳, 머리말, FAQ Q3 산문, FAQ Q8 표, README 2곳, 부록 머리말.

```
$ grep -c '2분 9초\|약 13초\|MLM 13초' 21_en_bert_classify/*.ipynb 21_en_bert_classify/README.md
0 / 0 / 0
```

### 재실행 검증

수정본을 T4 에서 두 판 돌렸습니다 — 각 `status: ok`, 에러 0, `source_sha256` 소스와 일치 (133.2초 / 132.2초). 시간 서술이 실측과 맞고, 이 과정에서 챕터의 다른 수치(파라미터 `11,103,290`, MLM perplexity 약 1,300, Ch 10 인용 `0.8980`, accuracy 격차 20%p 초과)도 함께 확인했습니다.

`pages/` 재변환과 tex 재생성은 #45 에서 이어가고, 이 이슈는 닫습니다.
